// @subql/types/dist/project.js
"use strict";
// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubqlHandlerKind = exports.SubqlDatasourceKind = void 0;
// export enum SubqlKind {
//   Runtime = 'substrate/Runtime',
//   BlockHandler = 'substrate/BlockHandler',
//   CallHandler = 'substrate/CallHandler',
//   EventHandler = 'substrate/EventHandler',
// }
var SubqlDatasourceKind;
(function (SubqlDatasourceKind) {
    SubqlDatasourceKind["Runtime"] = "substrate/Runtime";
    // Custom = 'substrate/Custom',
})(SubqlDatasourceKind = exports.SubqlDatasourceKind || (exports.SubqlDatasourceKind = {}));
var SubqlHandlerKind;
(function (SubqlHandlerKind) {
    SubqlHandlerKind["Block"] = "substrate/BlockHandler";
    SubqlHandlerKind["Call"] = "substrate/CallHandler";
    SubqlHandlerKind["Event"] = "substrate/EventHandler";
})(SubqlHandlerKind = exports.SubqlHandlerKind || (exports.SubqlHandlerKind = {}));
