// @polkadot/types/metadata/decorate/storage/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.decorateStorage = decorateStorage;

var _util = require("@polkadot/util");

var _createFunction = require("./createFunction.cjs");

var _getStorage = require("./getStorage.cjs");

var _util2 = require("./util.cjs");

// Copyright 2017-2021 @polkadot/types authors & contributors
// SPDX-License-Identifier: Apache-2.0

/** @internal */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function decorateStorage(registry, {
  pallets
}, _metaVersion) {
  const result = (0, _getStorage.getStorage)(registry);

  for (let p = 0; p < pallets.length; p++) {
    const {
      name,
      storage
    } = pallets[p];

    if (storage.isSome) {
      const section = (0, _util.stringCamelCase)(name);
      const {
        items,
        prefix: _prefix
      } = storage.unwrap();

      const prefix = _prefix.toString();

      const newModule = {
        palletVersion: (0, _util2.createRuntimeFunction)({
          method: 'palletVersion',
          prefix,
          section
        }, (0, _createFunction.createKeyRaw)(registry, {
          method: ':__STORAGE_VERSION__:',
          prefix: name.toString()
        }, [], [], []), {
          docs: 'Returns the current pallet version from storage',
          type: 'u16'
        })(registry)
      };

      for (let i = 0; i < items.length; i++) {
        const meta = items[i];
        const method = meta.name.toString(); // For access, we change the index names, i.e. System.Account -> system.account

        newModule[(0, _util.stringLowerFirst)(method)] = (0, _createFunction.createFunction)(registry, {
          meta,
          method,
          prefix,
          section
        }, {});
      }

      result[section] = newModule;
    }
  }

  return result;
}
